text/010-wasm-pack-watch-and-serve.md

```diff
@@ -0,0 +1,230 @@
+                           generating this TypeScript file.
+    -h, --help             Prints help information
+        --profiling        Create a profiling build. Enable optimizations and debug info.
+        --release          Create a release build. Enable optimizations and disable debug info.
+    -V, --version          Prints version information
+        --no-watch         Do not watch the crate for changes to automaticaly rebuild.
+
+OPTIONS:
+    -m, --mode <mode>          Sets steps to be run. [possible values: no-install, normal, force] [default: normal]
+    -d, --out-dir <out_dir>    Sets the output directory with a relative path. [default: pkg]
+    -s, --scope <scope>        The npm scope to use in package.json, if any.
+    -t, --target <target>      Sets the target environment. [possible values: browser, nodejs, no-modules] [default:
+                               browser]
+    -p, --port <port>          Bind to this port number with the local development HTTP server. [default: 8000]
+
+ARGS:
+    <path>                The path to the Rust crate.
+    <extra_options>...    List of extra options to pass to `cargo build`
+```
+
+### Implementation
+
+Rather than integrating an HTTP server into the `wasm-pack` binary itself, we
+should leverage its ability to download and run other binaries. Exactly *which*
+local HTTP server binary is left as an unresolved question.
+
+The subcommand will start the local HTTP server and then execute the `watch`
+subcommand (unless `--no-watch` is supplied).
+
+# Drawbacks, Rationale, and Alternatives
+
+The primary drawbacks are authoring and maintaining two more subcommands to
+implement functionality that is already available in external, third-party
+tools. We try to mitigate this downside by using existing local HTTP server
+binaries rather than building an HTTP server directly into `wasm-pack` itself.
+
+## Alternative: Build the HTTP Server into `wasm-pack`
+
+An alternative design would be to build the HTTP server into `wasm-pack` itself,
+using one of the existing crates for building Web servers like rocket, actix,
+hyper, tide, etc:
+
+* **Pros:**
+  * The `wasm-pack serve` command is ready to roll immediately after `wasm-pack`
+    is installed, and doesn't need to hit the network the first time you run
+    it. This situation only arises when folks who already downloaded `wasm-pack`
+    do their first build/serve offline, disconnected from the internet. This
+    seems like a niche enough situation that we can take the trade off.
+* **Cons:**
+  * Building the HTTP server into the `wasm-pack` binary is less robust: by
+    shelling out to an external binary for our local server, we get process
+    isolation, which makes error recovery simpler.
+  * More work to implement. `wasm-pack` already has plenty of infrastructure for
+    downloading and working with external binaries, so using external HTTP
+    servers should be fairly easy to implement.
+
```

Comment:
Thinking about this, one feature of `webpack-dev-server` is that when you request a page it actually waits to send any data until the current build is done. In that sense I think that a con of using an external binary is that we can't implement logic like that.

For example if I make a change and then quickly go and refresh a web page, I might get stale output if the build hasn't finished yet.

I wonder though if we could perhaps destroy `--out-dir` on each build to ensure that if the page is loaded stale artifacts are never served up? (not a great experience but perhaps better than seeing stale data)